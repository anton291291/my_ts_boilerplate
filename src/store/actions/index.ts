export const PlayerActions = {
    setIsPlay: () =>
        ({
            type: 'PLAYER:PLAY',
            payload: true
        } as const),

    setIsStop: () =>
        ({
            type: 'PLAYER:STOP',
            payload: false
        } as const),

    setReset: () =>
        ({
            type: 'PLAYER:RESET'
        } as const),

    setRandomIndex: (index: number) =>
        ({
            type: 'PLAYER:SET_RANDOM_INDEX',
            payload: index
        } as const),

    setSpeed: (index: number) =>
        ({
            type: 'PLAYER:SET_SPEED_INDEX',
            payload: index
        } as const),

    setGen: (value: number) =>
        ({
            type: 'PLAYER:SET_GENERATION',
            payload: value
        } as const),
    getName: () =>
        ({
            type: 'PLAYER:GET_NAME'
        } as const)
};

export const GreetingFormActions = {
    setStart: () =>
        ({
            type: 'GREETINGFORM:START'
        } as const)
};

export const AxisActions = {
    increaseXaxis: (value: number) =>
        ({
            type: 'AXIS:INCREASE_X',
            payload: value
        } as const),
    decreaseXaxis: (value: number) =>
        ({
            type: 'AXIS:DECREASE_X',
            payload: value
        } as const),
    increaseYaxis: (value: number) =>
        ({
            type: 'AXIS:INCREASE_Y',
            payload: value
        } as const),
    decreaseYaxis: (value: number) =>
        ({
            type: 'AXIS:DECREASE_Y',
            payload: value
        } as const)
};

export const IntervalIDAction = {
    setIntervalID: (id: number) =>
        ({
            type: 'INTERVALID:SET_VALUE',
            payload: id
        } as const)
};

export const CellsActions = {
    clickCell: (index: number) =>
        ({
            type: 'CELLS:CLICK',
            payload: index
        } as const),
    randomCells: () =>
        ({
            type: 'CELLS:RANDOM'
        } as const),

    simulateLife: () =>
        ({
            type: 'CELLS:SIMULATE_LIFE'
        } as const)
};
