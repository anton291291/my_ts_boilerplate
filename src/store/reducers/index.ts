import { ActionTypes } from '../actions/types';
import { simulateLife } from '@/utills/algorithm';
import { randomCells } from '@/utills/helper';

const y = 15;
const x = 30;

const initialState: StateType = {
    axis: { x: x, y: y },
    cells: Array.from({ length: y * x }, (item, index) => ({
        index: index,
        isClicked: false
    })),
    gen: 1,
    name: localStorage.getItem('name') || null,
    intervalID: null,
    isPlay: false,
    speed: 1,
    randomIndex: 0.7
};

export const GridReducer = (
    state = initialState,
    action: ActionTypes
): StateType => {
    const x = state.axis.x;
    const y = state.axis.y;
    let length = x * y - 1;

    switch (action.type) {
        case 'PLAYER:PLAY':
            return { ...state, isPlay: action.payload };
        case 'PLAYER:STOP':
            return { ...state, isPlay: action.payload };
        case 'PLAYER:RESET':
            return {
                ...state,
                cells: state.cells.map((obj) => {
                    return { ...obj, isClicked: false };
                }),
                isPlay: false,
                gen: 1
            };
        case 'PLAYER:SET_GENERATION':
            return { ...state, gen: action.payload };
        case 'PLAYER:SET_RANDOM_INDEX':
            return { ...state, randomIndex: action.payload };
        case 'PLAYER:SET_SPEED_INDEX':
            return { ...state, speed: action.payload };
        case 'AXIS:DECREASE_Y':
            return {
                ...state,
                axis: { ...state.axis, y: action.payload },
                cells: state.cells.slice(0, action.payload * x)
            };
        case 'AXIS:INCREASE_Y':
            return {
                ...state,
                axis: { ...state.axis, y: action.payload },
                cells: [
                    ...state.cells,
                    ...Array.from(
                        { length: action.payload * x - x * y },
                        () => ({
                            index: ++length,
                            isClicked: false
                        })
                    )
                ]
            };
        case 'AXIS:DECREASE_X':
            return {
                ...state,
                axis: { ...state.axis, x: action.payload },
                cells: state.cells.slice(0, action.payload * y)
            };
        case 'AXIS:INCREASE_X':
            return {
                ...state,
                axis: { ...state.axis, x: action.payload },
                cells: [
                    ...state.cells,
                    ...Array.from(
                        { length: action.payload * y - x * y },
                        () => ({
                            index: ++length,
                            isClicked: false
                        })
                    )
                ]
            };

        case 'INTERVALID:SET_VALUE':
            return { ...state, intervalID: action.payload };
        case 'CELLS:RANDOM':
            return {
                ...state,
                cells: state.cells.map((obj) => randomCells(obj, state))
            };
        case 'CELLS:CLICK':
            return {
                ...state,
                cells: state.cells.map((obj) => {
                    if (obj.index === action.payload)
                        return { ...obj, isClicked: !obj.isClicked };
                    return obj;
                })
            };
        case 'CELLS:SIMULATE_LIFE':
            return {
                ...state,
                cells: state.cells.map((obj, index, arr) =>
                    simulateLife(obj, index, arr, state)
                ),
                gen: state.gen + 1
            };
        case 'GREETINGFORM:SET_NAME':
            return { ...state, name: localStorage.getItem('name') };

        case 'GREETINGFORM:START':
            return {
                ...state,
                cells: state.cells.map((obj) => {
                    if (Math.random() > 0.7) {
                        return { ...obj, isClicked: true };
                    }
                    return obj;
                }),
                isPlay: true
            };
        default:
            return state;
    }
};
