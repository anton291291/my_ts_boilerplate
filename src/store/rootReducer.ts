import { combineReducers } from 'redux';

import { reducer as formReducer } from 'redux-form';
import { GridReducer } from './reducers/index';

export const rootReducer = combineReducers({
    form: formReducer,
    grid: GridReducer
});

export type RootState = ReturnType<typeof rootReducer>;

