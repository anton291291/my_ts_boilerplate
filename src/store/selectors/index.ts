import { RootState } from '@/store/rootReducer';

export const speedSelector = (state: RootState) => state.grid.speed;
export const isPlaySelector = (state: RootState) => state.grid.isPlay