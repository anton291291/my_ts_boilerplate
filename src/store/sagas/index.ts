import { select, put, takeLatest, delay } from 'redux-saga/effects';
import { speedSelector, isPlaySelector } from '../selectors';
import { PlayerActions, GreetingFormActions, CellsActions } from '../actions';

export function* startSimulateLife() {
    const speed = yield select(speedSelector);
    while (yield select(isPlaySelector)) {
        yield put(CellsActions.simulateLife());
        yield delay(speed * 100);
    }
}

export function* startGameFromForm() {
    yield put(GreetingFormActions.setReady());
    yield startSimulateLife();
}

export function* watchStartGame() {
    yield takeLatest(PlayerActions.setIsPlay().type, startSimulateLife);
}

export function* watchStartGameFromForm() {
    yield takeLatest(GreetingFormActions.setStart().type, startGameFromForm);
}
