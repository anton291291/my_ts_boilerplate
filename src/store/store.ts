import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import createSagaMiddleware from 'redux-saga'

import {rootReducer} from './rootReducer';
import { rootSaga } from './sagas/rootSaga';

declare global {
    interface Window {
        __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
    }
}

const composeEnhancers = this?.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const sagaMiddleware = createSagaMiddleware()

const middleware = composeEnhancers(applyMiddleware(thunk, sagaMiddleware));

const store = createStore(rootReducer, middleware);

sagaMiddleware.run(rootSaga)

export default store;
