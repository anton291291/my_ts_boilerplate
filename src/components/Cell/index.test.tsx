import React from 'react';
import { mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';

import { Cell } from './index';
import store from '@/store/store';
import { Provider } from 'react-redux';

describe('Cell', () => {
    const output = mount(
        <Provider store={store}>
            <Cell index={0} isClicked={false} />
        </Provider>
    );

    test('should render correctly', () => {
        expect(mountToJson(output)).toMatchSnapshot();
    });
});
