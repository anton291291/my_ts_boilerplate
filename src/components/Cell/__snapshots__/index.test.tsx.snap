// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Cell
    index={0}
    isClicked={false}
  >
    <StyledCell
      isClicked={false}
      onClick={[Function]}
    >
      <div
        className="StyledCell-ybaiwl-0 cTvkpX"
        onClick={[Function]}
      />
    </StyledCell>
  </Cell>
</Provider>
`;
